### PR TITLE
Fix Issue 21678 - "_d_arraysetlengthT is not callable using argument …

### PR DESCRIPTION
**Pull request by BorisCarvajal**
…types" on chained array length assignment

### CODE REVIEW

File: test/compilable/vcg-ast-arraylength.d [hunk 1]
```diff
@@ -12,4 +12,12 @@ void main()
...
 	// should convert to arr = arr[0 .. 0];
 	arr.length = 0;
+
+    // https://issues.dlang.org/show_bug.cgi?id=21678
+    int[] f;
+    int[] a;
+
+    a.length = f.length = 0;
+    const x = 0;
+    a.length = f.length = x;
```
Comment (MoonlightSentinel) [inline test/compilable/vcg-ast-arraylength.d:22]: Add a check to verify that the type of the expression is size_t, e.g.

`static assert(is(typeof(a.length = 0) == size_t)));`

File: src/dmd/sideeffect.d [hunk 2]
```diff
@@ -332,16 +332,10 @@ bool discardValue(Expression e)
...
-            /* Check for compiler-generated code of the form  auto __tmp, e, __tmp;
-             * In such cases, only check e for side effect (it's OK for __tmp to have
-             * no side effect).
-             * See https://issues.dlang.org/show_bug.cgi?id=4231 for discussion
-             */
-            auto fc = firstComma(ce);
-            if (fc.op == TOK.declaration && ce.e2.op == TOK.variable && (cast(DeclarationExp)fc).declaration == (cast(VarExp)ce.e2).var)
-            {
+            // Don't complain about compiler-generated comma expressions
+            if (ce.isGenerated)
```
Comment (MoonlightSentinel) [inline src/dmd/sideeffect.d:336]: Maybe add a sanity check that at least one of the expressions has a side effect, e.g. as

`assert(hasSideEffect(ce.e1) || hasSideEffect(ce.e2))`

Dmd shouldn't create `CommaExp` without any side effect.

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -9240,13 +9240,16 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             if (!verifyHookExist(exp.loc, *sc, Id._d_arraysetlengthTImpl, "resizing arrays"))
                 return setError();
 
+            auto lc = lastComma(exp.e2);
+            lc = lc.optimize(WANTvalue);
```
Comment (MoonlightSentinel) [inline src/dmd/expressionsem.d:9244]: Why is this necessary?

**Comment by RazvanN7** [comment]
The issue is a regression and therefore this PR should target stable.

File: src/dmd/sideeffect.d (hunk 2 above)
Comment (BorisCarvajal) [reply src/dmd/sideeffect.d:336]: Good idea.

File: src/dmd/expressionsem.d (hunk 3 above)
Comment (BorisCarvajal) [reply src/dmd/expressionsem.d:9244]: 1. To handle a chain `a.length = f.length = 0;` -> `(a = a[0..0] , (f = f[0..0] , 0));`
2. `optimize()` is needed for non-trivial cases:
```
    const x = 0;
    a.length = f.length = x;
```

File: test/compilable/vcg-ast-arraylength.d (hunk 1 above)
Comment (BorisCarvajal) [reply test/compilable/vcg-ast-arraylength.d:22]: Test added.
I had to force the type to the same as the length, good catch.

File: src/dmd/sideeffect.d (hunk 2 above)
Comment (BorisCarvajal) [reply src/dmd/sideeffect.d:336]: Added the proposed `assert`.

Comment (BorisCarvajal) [reply src/dmd/sideeffect.d:336]: The assert has revealed that DMD creates a chain of commas for struct/class `invariant`, as these are functions with no side-effects the condition is not met.
I'm back to returning a plain `false` since the old deleted condition was just a pattern to detect the most common generated comma from a time where commas were allowed everywhere.

**Review by thewilsonator — APPROVED** [review]
